app/src/main/java/ch/epfl/culturequest/ui/scan/CameraSetup.java

```diff
@@ -0,0 +1,341 @@
+
+        // This will make the camera frame fill the texture view, if you'd like to fit it into the view swap the "<" sign for ">"
+        if (viewAspectRatio < imageAspectRatio) {
+            // If the view is "thinner" than the image constrain the height and calculate the scale for the texture width
+            scale = new PointF((viewRect.height() / viewRect.width()) * ((float) height / (float) width), 1f);
+        } else {
+            scale = new PointF(1f, (viewRect.width() / viewRect.height()) * ((float) width / (float) height));
+        }
+
+        if (rotationDegrees % 180 != 0) {
+            // If we need to rotate the texture 90º we need to adjust the scale
+            float multiplier = viewAspectRatio < imageAspectRatio ? w/h : h/w;
+            scale.x *= multiplier;
+            scale.y *= multiplier;
+        }
+
+        Matrix matrix = new Matrix();
+        // Set the scale
+        matrix.setScale(scale.x, scale.y, viewRect.centerX(), viewRect.centerY());
+        if (rotationDegrees != 0) {
+            // Set rotation of the device isn't upright
+            matrix.postRotate(0 - rotationDegrees, viewRect.centerX(), viewRect.centerY());
+        }
+        // Transform the texture
+        textureView.setTransform(matrix);
+    }
+
+    //Returns the rotation of the display
+    private int getDisplayRotation() {
+        switch (textureView.getDisplay().getRotation()) {
+            case Surface.ROTATION_0:
+            default:
+                return 0;
+            case Surface.ROTATION_90:
+                return  90;
+            case Surface.ROTATION_180:
+                return  180;
+            case Surface.ROTATION_270:
+                return 270;
+        }
+    }
+
+    /**
+     * Get the preview size from the camera characteristics
+     * @param characteristics the characteristics of the camera
+     * @return the preview size
+     */
+    private Size getPreviewSize(CameraCharacteristics characteristics) {
+        return Arrays.stream(characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP).getOutputSizes(ImageFormat.JPEG)).max(Comparator.comparingInt(x -> x.getHeight()*x.getWidth())).get();
+    }
+
+    /**
+     * Get the sensor orientation of the camera
+     * @param characteristics the characteristics of the camera
+     * @return the sensor orientation
+     */
+    private int getCameraSensorOrientation(CameraCharacteristics characteristics) {
+        Integer cameraOrientation = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
+        return (360 - (cameraOrientation != null ? cameraOrientation : 0)) % 360;
+    }
+
+    ///////////////////////////// Image Reader ///////////////////////////
+    private final ImageReader.OnImageAvailableListener onImageAvailableListener = imageReader -> {
+        Image image = imageReader.acquireLatestImage();
+        if (image != null) {
+            latestImage = image;
+        }
+    };
+
```

Comment:
As this "onImageAvailableListener" is a class attribute, maybe it could be easier to see that it exists if you put if before the constructor of the class.

Reply:
Yes you are right, I will do so!